multicluster/controllers/multicluster/clusterset_controller.go

```diff
@@ -58,5 +117,172 @@ func (r *ClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+			wellKnownClusterClaimIDExist = true
+			clusterClaimID = clusterClaim.Value
+		}
+	}
+
+	if !wellKnownClusterSetClaimIDExist {
+		return fmt.Errorf("ClaimClaim not configured for Name=%s", multiclusterv1alpha1.WellKnownClusterClaimClusterSet)
+	}
+
+	if !wellKnownClusterClaimIDExist {
+		return fmt.Errorf("ClaimClaim not configured for Name=%s", multiclusterv1alpha1.WellKnownClusterClaimID)
+	}
+
+	configExists := false
+	if r.IsLeader {
+		// Validate the Namespace is the same
+		if clusterSet.Spec.Namespace != clusterSet.GetNamespace() {
+			return fmt.Errorf("ClusterSet Namespace " + clusterSet.Spec.Namespace + " is different from " +
+				clusterSet.GetNamespace())
+		}
+		for _, leader := range clusterSet.Spec.Leaders {
+			if clusterClaimID == leader.ClusterID {
+				configExists = true
+				break
+			}
+		}
+		if !configExists {
+			return fmt.Errorf("local cluster is not defined as Leader in the ClusterSet")
+		}
+	} else {
+		for _, member := range clusterSet.Spec.Members {
+			if clusterClaimID == member.ClusterID {
+				configExists = true
+				break
+			}
+		}
+		if !configExists {
+			return fmt.Errorf("local cluster not defined as Member in ClusterSet")
+		}
+	}
+
+	r.clusterID = common.ClusterID(clusterClaimID)
+	r.clusterSetID = common.ClusterSetID(clusterSetClaimID)
+	r.clusterSetConfig = clusterSet.DeepCopy()
+
+	return nil
+}
+
+func (r *ClusterSetReconciler) createLocalCommonArea(clusterSet *multiclusterv1alpha1.ClusterSet) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if r.LocalCommonArea == nil {
+		r.LocalCommonArea = core.NewLocalCommonArea(r.Client, r.clusterID, clusterSet.GetNamespace())
+	}
+}
+
+func (r *ClusterSetReconciler) updateMultiClusterSetOnMemberCluster(clusterSet *multiclusterv1alpha1.ClusterSet) error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if r.RemoteCommonAreaManager == nil {
+		r.RemoteCommonAreaManager = core.NewRemoteCommonAreaManager(r.clusterSetID, r.clusterID)
+		go func() {
+			err := r.RemoteCommonAreaManager.Start()
+			if err != nil {
+				klog.ErrorS(err, "error starting RemoteCommonAreaManager")
+				r.mutex.Lock()
+				r.RemoteCommonAreaManager = nil
+				r.clusterSetID = common.InvalidClusterSetID
+				r.clusterID = common.InvalidClusterID
+				r.clusterSetConfig = nil
+				r.mutex.Unlock()
+			}
+		}()
+	}
+
+	currentLeaders := r.RemoteCommonAreaManager.GetRemoteCommonAreas()
+	newLeaders := clusterSet.Spec.Leaders
+
+	var addedLeaders []*multiclusterv1alpha1.MemberCluster
+	var removedLeaders map[common.ClusterID]core.RemoteCommonArea
+
+	for _, leader := range newLeaders {
+		leaderID := common.ClusterID(leader.ClusterID)
+		_, found := currentLeaders[leaderID]
+		if !found {
+			addedLeaders = append(addedLeaders, leader.DeepCopy())
+		} else {
+			// In the end currentLeaders will only have removed leaders
+			delete(currentLeaders, leaderID)
+		}
+	}
+
+	klog.InfoS("ClusterSet update", "addedLeaders", addedLeaders, "removedLeaders", removedLeaders)
+
+	var multiErr error
```

Comment:
I feel we should unlock r.mutex before these blocking calls. Do you see an issue for that?

Reply:
If I remove the goroutine, I think we dont need the lock, otherwise there can be null-pointer if r.RemoteCommonAreaManager gets set to nil from the goroutine. Ill remove the lock when I remove the go routine in another PR?

Reply:
I actually removed the goroutine in the same PR, I also removed the locks. But needed to add it back for status implementation in the status implementation PR. Since status computation runs on a timer in a separate goroutine and we need to synchronize with Reconcile handling.